yt/frontends/enzo_p/data_structures.py

```diff
@@ -0,0 +1,385 @@
+                    grid_id = rbid
+                    parent_id = -1
+                else:
+                    grid_id = child_id
+                    # Try the last parent_id first
+                    if last_pid is not None and \
+                      is_parent(self.grids[last_pid].block_name, block_name):
+                        parent_id = last_pid
+                    else:
+                        parent_id = self.grids[rbid].get_parent_id(block_name)
+                    last_pid = parent_id
+                    child_id += 1
+
+                my_grid = self.grid(
+                    grid_id, self, block_name,
+                    filename=os.path.join(self.directory, block_file))
+                my_grid.Level = level
+                my_grid._parent_id = parent_id
+
+                self.grids[grid_id]           = my_grid
+                self.grid_levels[grid_id]     = level
+                self.grid_left_edge[grid_id]  = left
+                self.grid_right_edge[grid_id] = right
+                self.grid_dimensions[grid_id] = self.ds.active_grid_dimensions
+
+                if level > 0:
+                    self.grids[parent_id].add_child(my_grid)
+
+                bnl = nnl + 1
+                pbar.update(1)
+            lstr = buff[bnl:]
+            offset += fblock
+
+        f.close()
+        pbar.finish()
+
+        slope = self.ds.domain_width / \
+          self.ds.arr(np.ones(3), "code_length")
+        self.grid_left_edge   = self.grid_left_edge  * slope + \
+          self.ds.domain_left_edge
+        self.grid_right_edge  = self.grid_right_edge * slope + \
+          self.ds.domain_left_edge
+
+    def _populate_grid_objects(self):
+        for g in self.grids:
+            g._prepare_grid()
+            g._setup_dx()
+        self.max_level = self.grid_levels.max()
+
+    def _setup_derived_fields(self):
+        super(EnzoPHierarchy, self)._setup_derived_fields()
+        for fname, field in self.ds.field_info.items():
+            if not field.particle_type: continue
+            if isinstance(fname, tuple): continue
+            if field._function is NullFunc: continue
+
+    def _detect_output_fields(self):
+        self.field_list = []
+        # Do this only on the root processor to save disk work.
+        if self.comm.rank in (0, None):
+            mylog.info("Gathering a field list (this may take a moment.)")
+            field_list = set()
+            random_sample = self._generate_random_grids()
+            for grid in random_sample:
+                if not hasattr(grid, 'filename'): continue
+                try:
+                    gf = self.io._read_field_names(grid)
+                except self.io._read_exception:
+                    raise IOError("Grid %s is a bit funky?", grid.id)
+                mylog.debug("Grid %s has: %s", grid.id, gf)
+                field_list = field_list.union(gf)
+            if "AppendActiveParticleType" in self.dataset.parameters:
```

Comment:
Ok, yeah, this whole chunk I didn't bother to touch because we're still finalizing output of particle fields in Enzo-P.  I can clean it up, though.